allennlp/data/vocabulary.py

```diff
@@ -285,6 +287,40 @@ def from_dataset(cls,
                           max_vocab_size=max_vocab_size,
                           non_padded_namespaces=non_padded_namespaces)
 
+    @classmethod
+    def from_params(cls, params: Params, dataset=None):
+        """
+        There are two possible ways to build a vocabulary; from a
+        pre-existing dataset, using :func:`Vocabulary.from_dataset`, or
+        from a pre-saved vocabulary, using :func:`Vocabulary.from_files`.
+        This method wraps both of these options, allowing their specification
+        from a ``Params`` object, generated from a JSON configuration file.
+
+        Parameters
+        ----------
+        params: Params, required.
+        dataset: Dataset, optional.
+            If ``params`` doesn't contain a ``vocabulary_directory`` key,
+            the ``Vocabulary`` can be built directly from a ``Dataset``.
+
+        Returns
+        -------
+        A ``Vocabulary``.
+        """
+        vocabulary_directory = params.pop("vocabulary_directory", None)
+        if not vocabulary_directory and not dataset:
+            raise ConfigurationError("You must provide either a Params object containing a "
+                                     "vocab_directory key or a Dataset to build a vocabulary from.")
+        if vocabulary_directory and dataset:
+            logger.warning("from_params was passed both a vocabulary_directory and a Dataset. This "
+                           "behaviour is not well defined, so backing off to using the vocabulary "
+                           "defined at %s.", vocabulary_directory)
+
+        if vocabulary_directory:
+            params.assert_empty("Vocabulary - from files")
+            return Vocabulary.from_files(vocabulary_directory)
+        return Vocabulary.from_dataset(dataset, **params.as_dict())
```

Comment:
We've typically popped things off explicitly, except when passing the parameters off to pytorch functions, where we're wrapping things with different APIs and _can't_ explicitly pop off parameters.  We give better error messages when things are popped off, and you can call `params.assert_empty()`.